### PR TITLE
Convert evos, prevo, cosmeticFormes, and otherFormes from IDs to names

### PR DESCRIPTION
**Pull request by KrisXV**

### CODE REVIEW

**Comment by Zarel** [comment]
Let's actually validate the names. Like how I do it for factory sets.

File: sim/dex.ts [hunk 1]
```diff
@@ -337,7 +337,7 @@ export class ModdedDex {
 	getForme(speciesid: string | Species): string {
 		const id = toID(speciesid || '');
 		const species = this.getSpecies(id);
-		if (species.cosmeticFormes && species.cosmeticFormes.includes(id)) {
+		if (species.cosmeticFormes?.map(toID).includes(id)) {
```
Comment (Zarel) [inline sim/dex.ts:340]:
```suggestion
		if (species.cosmeticFormes?.includes(species.name)) {
```

Comment (Zarel) [reply sim/dex.ts:340]: Wait a minute what is this function even doing? This does exactly the same thing as `getSpecies`, doesn't it?

Comment (KrisXV) [reply sim/dex.ts:340]: This recommended change makes tests fail.

```diff
1) Dex#getForme
       should handle Flabébé formes:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'Flabébé'
- 'Flabébé-Yellow'
            ^
      + expected - actual

      -Flabébé
      +Flabébé-Yellow

      at Context.<anonymous> (test\sim\dex.js:37:10)
      at processImmediate (internal/timers.js:439:21)
      at process.topLevelDomainCallback (domain.js:126:23)
```

Comment (Zarel) [reply sim/dex.ts:340]: Apparently this is a client-only feature

> Dex.getSpecies('flabebeyellow').name
> "Flabébé-Yellow"

The server should probably be updated to match client.

Comment (Zarel) [reply sim/dex.ts:340]: Although feel free to leave this alone and do the update in a different commit, I guess.

**Comment by Zarel** [comment]
(I _am_ going to block the merge on validating the names, though. But feel free to tell me to do it if you're having trouble with that.)

File: sim/dex.ts (hunk 1 above)
Comment (scheibo) [reply sim/dex.ts:340]: > This recommended change makes tests fail.

Glad I added the test

File: test/sim/pokedex.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+const assert = require('./../assert');
+
+describe('Pokedex', function () {
+	it('should make sure every Pokemon name is correct', function () {
+		for (const pokemonid in Dex.data.Pokedex) {
+			const name = Dex.data.Pokedex[pokemonid].name;
+			const id = toID(name);
+			assert(pokemonid === id, `Mismatched Pokemon key "${pokemonid}" of ${name}`);
```
Comment (Zarel) [inline test/sim/pokedex.js:10]:
```suggestion
		const Pokedex = Dex.data.Pokedex;
		for (const pokemonid inPokedex) {
			const entry = Pokedex[pokemonid];
			assert.equal(toID(entry.name), pokemonid, `Mismatched Pokemon key "${pokemonid}" of ${entry.name}`);
			if (entry.evos) {
				for (const evo of entry.evos) {
					assert.equal(evo, (Pokedex[toID(evo)] || {}).name, `Mismatched evo "${evo}" of ${entry.name}`);
				}
			}
```

(If we bump min version to Node 14, we can use `?.` in tests!)

Comment (Zarel) [reply test/sim/pokedex.js:10]: Anyway yeah, do this, but also for prevo, cosmeticFormes, otherFormes, and changesFrom.

File: test/sim/pokedex.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+const assert = require('./../assert');
+
+describe('Pokedex', function () {
+	it('should make sure every Pokemon name is correct', function () {
+		const Pokedex = Dex.data.Pokedex;
+		for (const pokemonid in Pokedex) {
+			const entry = Pokedex[pokemonid];
+			assert.equal(toID(entry.name), pokemonid, `Mismatched Pokemon key "${pokemonid}" of ${entry.name}`);
+			if (entry.prevo) {
+				assert.equal(entry.prevo, (Pokedex[toID(entry.prevo)] || {}).name, `Mismatched prevo "${entry.prevo}" of ${entry.name}`);
```
Comment (Zarel) [inline test/sim/pokedex.js:12]:
```suggestion
				const prevoEntry = Pokedex[toID(entry.prevo)] || {};
				assert.equal(entry.prevo, prevoEntry.name, `Mismatched prevo "${entry.prevo}" of ${entry.name}`);
				assert.notEqual(entry.num, prevoEntry.num, `Prevo ${entry.prevo} of ${entry.name} should have different dex number`);
				assert(prevoEntry.evos.includes(entry.name), `Prevo ${entry.prevo} of ${entry.name} should list it as an evo`);
```

Comment (Zarel) [reply test/sim/pokedex.js:12]: (but for all the others)

File: test/sim/pokedex.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+			}
+			if (entry.evos) {
+				for (const evo of entry.evos) {
+					assert.equal(evo, (Pokedex[toID(evo)] || {}).name, `Mismatched evo "${evo}" of ${entry.name}`);
```
Comment (Zarel) [inline test/sim/pokedex.js:16]:
```suggestion
					const evoEntry = Pokedex[toID(evo)] || {};
					assert.equal(evo, evoEntry.name, `Misspelled/nonexistent evo "${evo}" of ${entry.name}`);
					assert.notEqual(entry.num, evoEntry.num, `Evo ${entry.prevo} of ${entry.name} should have different dex number`);
					assert.equal(evoEntry.prevo, entry.name, `Evo ${entry.prevo} of ${entry.name} should list it as a prevo`);
```

File: test/sim/pokedex.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+				for (const forme of entry.otherFormes) {
+					const formeEntry = Pokedex[toID(forme)] || {};
+					assert.equal(forme, formeEntry.name, `Misspelled/nonexistent forme "${forme}" of ${entry.name}`);
+					assert.equal(entry.num, formeEntry.num, `Forme ${formeEntry.name} of ${entry.name} should have the same dex number`);
+				}
+			}
+			if (entry.battleOnly) {
+				const battleOnly = Array.isArray(entry.battleOnly) ? entry.battleOnly : [entry.battleOnly];
+				for (const battleForme of battleOnly) {
+					assert.equal(battleForme, (Pokedex[toID(battleForme)] || {}).name, `Misspelled/nonexistent battle-only forme "${battleForme}" of ${entry.name}`);
```
Comment (Zarel) [inline test/sim/pokedex.js:35]: check for same dex number?

File: test/sim/pokedex.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+				}
+			}
+			if (entry.changesFrom) {
+				assert.equal(entry.changesFrom, (Pokedex[toID(entry.changesFrom)] || {}).name, `Mismatched changesFrom value "${entry.changesFrom}" of ${entry.name}`);
```
Comment (Zarel) [inline test/sim/pokedex.js:39]: check for same dex number?

File: test/sim/pokedex.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+			}
```
Comment (Zarel) [inline test/sim/pokedex.js:40]: check that `cosmeticFormes` starts with the same name?

File: test/sim/pokedex.js (hunk 5 above)
Comment (KrisXV) [reply test/sim/pokedex.js:35]: This is already handled in otherFormes

File: test/sim/pokedex.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
+'use strict';
+
+const assert = require('./../assert');
+
+describe('Pokedex', function () {
+	it('should make sure every Pokemon name is correct', function () {
+		const Pokedex = Dex.data.Pokedex;
+		for (const pokemonid in Pokedex) {
+			const entry = Pokedex[pokemonid];
+			assert.equal(toID(entry.name), pokemonid, `Mismatched Pokemon key "${pokemonid}" of ${entry.name}`);
```
Comment (Zarel) [inline test/sim/pokedex.js:10]:
```suggestion

						assert.equal(toID(entry.name), pokemonid, `Mismatched Pokemon key "${pokemonid}" of ${entry.name}`);
						assert.equal(entry.name, entry.name.trim(), `Pokémon name "${entry.name}" should not start or end with whitespace`);
						assert(!entry.startsWith('-') && !entry.startsWith('-'), `Pokémon name "${entry.name}" should not start or end with a dash`);

```

File: test/sim/pokedex.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+			}
+			if (entry.changesFrom) {
+				const formeEntry = Pokedex[toID(entry.changesFrom)] || {};
+				assert.equal(entry.changesFrom, formeEntry.name, `Misspelled/nonexistent changesFrom value "${entry.changesFrom}" of ${entry.name}`);
+				assert.equal(entry.num, formeEntry.num, `changesFrom value ${formeEntry.name} of ${entry.name} should have the same dex number`);
+			}
+			if (entry.cosmeticFormes) {
+				for (const forme of entry.cosmeticFormes) {
+					assert(forme.startsWith(`${entry.name}-`), `Misspelled/nonexistent beginning of cosmetic forme name "${forme}" of ${entry.name}`);
+					assert(!forme.startsWith("-") && !forme.endsWith("-"), `Cosmetic forme name "${forme}" of ${entry.name} should not start or end with a hyphen`);
```
Comment (Zarel) [inline test/sim/pokedex.js:50]:
```suggestion
					assert(!forme.endsWith("-"), `Cosmetic forme name "${forme}" of ${entry.name} should not end with a hyphen`);
```

(Start is already checked on line 11.)

**Comment by Zarel** [comment]
Cool, neat, this is what I wanted.

File: test/sim/pokedex.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+		const Pokedex = Dex.data.Pokedex;
+		for (const pokemonid in Pokedex) {
+			const entry = Pokedex[pokemonid];
+			assert.equal(toID(entry.name), pokemonid, `Mismatched Pokemon key "${pokemonid}" of ${entry.name}`);
+			assert(!entry.name.startsWith("-") && !entry.name.endsWith("-"), `Pokemon name "${entry.name}" should not start or end with a hyphen`);
+			assert.equal(entry.name, entry.name.trim(), `Pokemon name "${entry.name}" should not start or end with whitespace`);
+			if (entry.prevo) {
+				const prevoEntry = Pokedex[toID(entry.prevo)] || {};
+				assert.equal(entry.prevo, prevoEntry.name, `Misspelled/nonexistent prevo "${entry.prevo}" of ${entry.name}`);
+				assert.notEqual(entry.num, prevoEntry.num, `Prevo ${entry.prevo} of ${entry.name} should have a different dex number`);
```
Comment (Zarel) [inline test/sim/pokedex.js:16]:
```suggestion
```

In hindsight, these are redundant and already checked in `entry.evos`

File: test/sim/pokedex.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+const assert = require('./../assert');
+
+describe('Pokedex', function () {
+	it('should make sure every Pokemon name is correct', function () {
+		const Pokedex = Dex.data.Pokedex;
+		for (const pokemonid in Pokedex) {
+			const entry = Pokedex[pokemonid];
+			assert.equal(toID(entry.name), pokemonid, `Mismatched Pokemon key "${pokemonid}" of ${entry.name}`);
+			assert(!entry.name.startsWith("-") && !entry.name.endsWith("-"), `Pokemon name "${entry.name}" should not start or end with a hyphen`);
+			assert.equal(entry.name, entry.name.trim(), `Pokemon name "${entry.name}" should not start or end with whitespace`);
```
Comment (Zarel) [inline test/sim/pokedex.js:12]:
```suggestion
			if (entry.baseSpecies) {
				const baseEntry = Pokedex[toID(entry.baseSpecies)] || {};
				assert(baseEntry.otherFormes.includes(entry.name), `Base species ${entry.baseSpecies} should have ${entry.name} listed as an otherForme`);
			}

```

File: test/sim/pokedex.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+					assert.equal(evo, evoEntry.name, `Misspelled/nonexistent evo "${evo}" of ${entry.name}`);
+					assert.notEqual(entry.num, evoEntry.num, `Evo ${evo} of ${entry.name} should have a different dex number`);
+					assert.equal(evoEntry.prevo, entry.name, `Evo ${evo} should have ${entry.name} listed as a prevo`);
+				}
+			}
+			if (entry.otherFormes) {
+				for (const forme of entry.otherFormes) {
+					const formeEntry = Pokedex[toID(forme)] || {};
+					assert.equal(forme, formeEntry.name, `Misspelled/nonexistent forme "${forme}" of ${entry.name}`);
+					assert.equal(entry.num, formeEntry.num, `Forme ${formeEntry.name} of ${entry.name} should have the same dex number`);
```
Comment (Zarel) [inline test/sim/pokedex.js:31]:
```suggestion
					assert.equal(formeEntry.baseSpecies, entry.name, `Forme ${forme} of ${entry.name} should have it as a baseSpecies`);
```

File: test/sim/pokedex.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+				}
+			}
+			if (entry.battleOnly) {
+				const battleOnly = Array.isArray(entry.battleOnly) ? entry.battleOnly : [entry.battleOnly];
+				for (const battleForme of battleOnly) {
+					const battleEntry = Pokedex[toID(battleForme)] || {};
+					assert.equal(battleForme, battleEntry.name, `Misspelled/nonexistent battle-only forme "${battleForme}" of ${entry.name}`);
+					assert.equal(entry.num, battleEntry.num, `Battle-only forme ${battleEntry.name} of ${entry.name} should have the same dex number`);
```
Comment (Zarel) [inline test/sim/pokedex.js:39]:
```suggestion
					assert.equal(battleForme, battleEntry.name, `Misspelled/nonexistent battle-only forme "${battleForme}" of ${entry.name}`);
					assert.equal(battleEntry.baseSpecies, entry.baseSpecies || entry.name, `Battle-only forme ${battleEntry.name} of ${entry.name} should have the same baseSpecies`);
```

Sharing a `baseSpecies` will eliminate the need to do the other checks (which are done in `entry.name`).

File: test/sim/pokedex.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+				}
+			}
+			if (entry.changesFrom) {
+				const formeEntry = Pokedex[toID(entry.changesFrom)] || {};
+				assert.equal(entry.changesFrom, formeEntry.name, `Misspelled/nonexistent changesFrom value "${entry.changesFrom}" of ${entry.name}`);
+				assert.equal(entry.num, formeEntry.num, `changesFrom value ${formeEntry.name} of ${entry.name} should have the same dex number`);
```
Comment (Zarel) [inline test/sim/pokedex.js:45]:
```suggestion
				assert.equal(formeEntry.baseSpecies, entry.baseSpecies || entry.name, `Original forme ${formeEntry.name} of ${entry.name} should have the same baseSpecies`);
```

File: test/sim/pokedex.js [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+					assert.equal(entry.num, formeEntry.num, `Forme ${formeEntry.name} of ${entry.name} should have the same dex number`);
+					assert.equal(formeEntry.baseSpecies, entry.name, `Forme ${forme} of ${entry.name} should have it as a baseSpecies`);
+				}
+			}
+			if (entry.battleOnly) {
+				const battleOnly = Array.isArray(entry.battleOnly) ? entry.battleOnly : [entry.battleOnly];
+				for (const battleForme of battleOnly) {
+					const battleEntry = Pokedex[toID(battleForme)] || {};
+					assert.equal(battleForme, battleEntry.name, `Misspelled/nonexistent battle-only forme "${battleForme}" of ${entry.name}`);
+					assert.equal(battleEntry.baseSpecies, entry.baseSpecies || entry.name, `Battle-only forme ${battleEntry.name} of ${entry.name} should have the same baseSpecies`);
```
Comment (Zarel) [inline test/sim/pokedex.js:43]:
```suggestion
					assert.equal(battleEntry.baseSpecies || battleEntry.name, entry.baseSpecies, `Battle-only forme ${entry.name} of ${battleEntry.name} should have the same baseSpecies`);
```

**Pull request merged**